README.md
# Vanilla Adaptive Maps
A way to include an Adaptive Map, based on [Brad Frost’s Adaptive Maps pattern](http://bradfrostweb.com/blog/post/adaptive-maps/), without requiring an external JavaScript library.

## Contents

Vanilla Adaptive Maps includes:

* `ChangeLog.md`. The list of changes to the core project.
* `README.md`. The file that you’re currently reading.
* `deploy.sh`. A script used for deploying the git repository into WordPress's SVN plugin repository. > ahem < this is not currently working properly.
* A `vanilla-adaptive-maps` subdirectory that contains the source code - a fully executable WordPress plugin.

## Features

Quick and easy shortcode to include, based on a street address:
* A link to a map of that street address,
* a static Google map of the street address for small screens, and
* an embedded full-featured Google map for larger screens.
* Ability to style maps and add custom marker for the address.

## Installation

Vanilla Adaptive Maps can be installed as follows:

1. Copy the `vanilla-adaptive-maps` directory into your `wp-content/plugins` directory.
2. In the WordPress dashboard, navigation to the *Plugins* page
Locate the menu item that reads “Vanilla Adaptive Maps.”
3. Click on *Activate.*

## Usage

To include an adaptive map, just use the shortcode `[vamap]` with your desired address.

E.g.: `[vamap addr="1203 Pearl St, Boulder, CO 80302"]`

## License

Vanilla Adaptive Maps is licensed under the GPL v2 or later.

> This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License, version 2, as published by the Free Software Foundation.

> This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

> You should have received a copy of the GNU General Public License along with this program; if not, write to the Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA

A copy of the license is included in the root of the plugin’s directory. The file is named `LICENSE.txt`.

# Credits

Vanilla Adaptive Maps was coded by [Jeremy Carlson](http://jeremycarlson.com/), based on JavaScript and the Adaptive Map pattern written by [Brad Frost](http://bradfrostweb.com/blog/post/adaptive-maps/).
The plugin structure is based on the [WordPress Plugin Boilerplate](https://github.com/ahmadawais/AA-Basic-Shortcode-Plugin-Boilerplate)


ChangeLog.md
# ChangeLog

* (January 26, 2017). 1.1 Add ability to style maps using JSON from snazzymaps.com.
* (September 22, 2015). 1.0.1 Plugin prepped for WordPress SVN.
* (August 18, 2015). Initial Release.


deploy.sh
#! /bin/bash
# A modification of Dean Clatworthy's deploy script as found here: https://github.com/deanc/wordpress-plugin-git-svn
# The difference is that this script lives in the plugin's git repo & doesn't require an existing SVN repo.

# main config
PLUGINSLUG="vanilla-adaptive-maps"
CURRENTDIR=`pwd`
MAINFILE="vanilla-adaptive-maps.php" # this should be the name of your main php file in the wordpress plugin

# git config
GITPATH="$CURRENTDIR/" # this file should be in the base of your git repository

# svn config
SVNPATH="/tmp/$PLUGINSLUG" # path to a temp SVN repo. No trailing slash required and don't add trunk.
SVNURL="http://plugins.svn.wordpress.org/vanilla-adaptive-maps/" # Remote SVN repo on wordpress.org, with no trailing slash
SVNUSER="jcdesign" # your svn username


# Let's begin...
echo ".........................................."
echo 
echo "Preparing to deploy wordpress plugin"
echo 
echo ".........................................."
echo 

# Check if subversion is installed before getting all worked up
if ! which svn >/dev/null; then
	echo "You'll need to install subversion before proceeding. Exiting....";
	exit 1;
fi

# Check version in readme.txt is the same as plugin file after translating both to unix line breaks to work around grep's failure to identify mac line breaks
NEWVERSION1=`grep "^Stable tag:" $GITPATH/readme.txt | awk -F' ' '{print $NF}'`
echo "readme.txt version: $NEWVERSION1"
NEWVERSION2=`grep "^Version:" $GITPATH/$MAINFILE | awk -F' ' '{print $NF}'`
echo "$MAINFILE version: $NEWVERSION2"

if [ "$NEWVERSION1" != "$NEWVERSION2" ]; then echo "Version in readme.txt & $MAINFILE don't match. Exiting...."; exit 1; fi

echo "Versions match in readme.txt and $MAINFILE. Let's proceed..."

if git show-ref --tags --quiet --verify -- "refs/tags/$NEWVERSION1"
	then 
		echo "Version $NEWVERSION1 already exists as git tag. Exiting...."; 
		exit 1; 
	else
		echo "Git version does not exist. Let's proceed..."
fi

cd $GITPATH
echo -e "Enter a commit message for this new version: \c"
read COMMITMSG
git commit -am "$COMMITMSG"

echo "Tagging new version in git"
git tag -a "$NEWVERSION1" -m "Tagging version $NEWVERSION1"

echo "Pushing latest commit to origin, with tags"
git push origin master
git push origin master --tags

echo 
echo "Creating local copy of SVN repo ..."
svn co $SVNURL $SVNPATH

echo "Clearing svn repo so we can overwrite it"
svn rm $SVNPATH/trunk/*

echo "Exporting the HEAD of master from git to the trunk of SVN"
git checkout-index -a -f --prefix=$SVNPATH/trunk/

echo "Ignoring github specific files and deployment script"
svn propset svn:ignore "deploy.sh
README.md
.git
.gitignore" "$SVNPATH/trunk/"

echo "Moving assets-wp-repo"
mkdir $SVNPATH/assets/
mv $SVNPATH/trunk/assets-wp-repo/* $SVNPATH/assets/
svn add $SVNPATH/assets/
svn delete $SVNPATH/trunk/assets-wp-repo

echo "Changing directory to SVN and committing to trunk"
cd $SVNPATH/trunk/
# Add all new files that are not set to be ignored
svn status | grep -v "^.[ \t]*\..*" | grep "^?" | awk '{print $2}' | xargs svn add
svn commit --username=$SVNUSER -m "$COMMITMSG"

echo "Creating new SVN tag & committing it"
cd $SVNPATH
svn copy trunk/ tags/$NEWVERSION1/
cd $SVNPATH/tags/$NEWVERSION1
svn commit --username=$SVNUSER -m "Tagging version $NEWVERSION1"

echo "Removing temporary directory $SVNPATH"
rm -fr $SVNPATH/

echo "*** FIN ***"

js/options.js
/* This, too, cribbed from 
    http://hughlashbrooke.com/2014/02/26/complete-versatile-options-page-class-wordpress-plugin/
*/
jQuery(document).ready(function($) {

    /***** Uploading images *****/

    var file_frame;

    jQuery.fn.uploadMediaFile = function( button, preview_media ) {
        var button_id = button.attr('id');
        var field_id = button_id.replace( '_button', '' );
        var preview_id = button_id.replace( '_button', '_preview' );

        // If the media frame already exists, reopen it.
        if ( file_frame ) {
          file_frame.open();
          return;
        }

        // Create the media frame.
        file_frame = wp.media.frames.file_frame = wp.media({
          title: jQuery( this ).data( 'uploader_title' ),
          button: {
            text: jQuery( this ).data( 'uploader_button_text' ),
          },
          multiple: false
        });

        // When an image is selected, run a callback.
        file_frame.on( 'select', function() {
          attachment = file_frame.state().get('selection').first().toJSON();
          jQuery("#"+field_id).val(attachment.id);
          if( preview_media ) {
            
            var url = ( undefined !== attachment.sizes.thumbnail ) ? attachment.sizes.thumbnail.url : attachment.url;
            jQuery("#"+preview_id).attr('src',url);
          }
        });

        // Finally, open the modal
        file_frame.open();
    }

    jQuery('.image_upload_button').click(function() {
        jQuery.fn.uploadMediaFile( jQuery(this), true );
    });

    jQuery('.image_delete_button').click(function() {
        jQuery(this).closest('td').find( '.image_data_field' ).val( '' );
        jQuery( '.image_preview' ).remove();
        return false;
    });
});

vamaps-options.php
<?php
class VAMSettingsPage
{
    /**
     * Holds the values to be used in the fields callbacks
     */
    private $options;

    /**
     * Start up
     */
    public function __construct()
    {
        add_action( 'admin_menu', array( $this, 'add_plugin_page' ) );
        add_action( 'admin_init', array( $this, 'page_init' ) );
    }

    /**
     * Add options page
     */
    public function add_plugin_page()
    {
        // This page will be under "Settings"
        $page = add_options_page(
            'Vanilla Adaptive Maps Settings', 
            'Vanilla Adaptive Maps Settings', 
            'manage_options', 
            'vamap-setting-admin', 
            array( $this, 'create_admin_page' )
        );
        
        add_action( 'admin_print_scripts-' . $page, array( $this, 'enqueue_scripts' ) );
    }
    
    /**
     * Register settings JS
     * @return void
     */
    public function register_scripts() {

    wp_register_script(
      'vam-admin-js', // script hook
      plugins_url( '/js/options.js' , __FILE__ ), // script location
      array(), // dependencies
      '1.0.0', // script version
      true // in footer?
    );
	}


    /**
     * Enqueue settings JS
     * @return void
     */
    public function enqueue_scripts() {

    // Include WP Media Scripts
    wp_enqueue_media();
    wp_enqueue_script( 'vam-admin-js' );
	}


    /**
     * Options page callback
     */
    public function create_admin_page()
    {
        // Set class property
        $this->options = get_option( 'vamap_style' );
        ?>
        <div class="wrap">
            <h2>Vanilla Adaptive Maps Settings</h2>           
            <form method="post" action="options.php">
            <?php
                // This prints out all hidden setting fields
                settings_fields( 'vamap_style_option_group' );   
                do_settings_sections( 'vamap-setting-admin' );
                submit_button(); 
            ?>
            </form>
        </div>
        <?php
    }

    /**
     * Register and add settings
     */
    public function page_init()
    {        
        register_setting(
            'vamap_style_option_group', // Option group
            'vamap_style', // Option name
            array( $this, 'sanitize' ) // Sanitize
        );

        add_settings_section(
            'vamap_style_settings', // ID
            'Style Settings', // Title
            array( $this, 'print_section_info' ), // Callback
            'vamap-setting-admin' // Page
        );  

        add_settings_field(
            'map_api_key', // ID
            '<a href="https://developers.google.com/maps/documentation/javascript/">Google API Key</a>', // Title 
            array( $this, 'map_api_callback' ), // Callback
            'vamap-setting-admin', // Page
            'vamap_style_settings' // Section           
        );      

        add_settings_field(
            'map_style', // ID
            'Map Style', // Title 
            array( $this, 'map_style_callback' ), // Callback
            'vamap-setting-admin', // Page
            'vamap_style_settings' // Section           
        );      

        add_settings_field(
            'map_icon', // ID
            'Marker Icon', // Title 
            array( $this, 'map_icon_callback' ), // Callback
            'vamap-setting-admin', // Page
            'vamap_style_settings' // Section           
        );

        VAMSettingsPage::register_scripts();
    }

    /**
     * Sanitize each setting field as needed
     *
     * @param array $input Contains all settings fields as array keys
     */
    public function sanitize( $input )
    {
        $new_input = array();
        if( isset( $input['map_api_key'] ) ) {
          $new_input['map_api_key'] = sanitize_text_field( $input['map_api_key'] );
        }

        if( isset( $input['map_style'] ) ) {
          $new_input['map_style'] = sanitize_text_field( $input['map_style'] );
        }

        if( isset( $input['map_icon'] ) ) {
          $new_input['map_icon'] = sanitize_text_field( $input['map_icon'] );
        }

        return $new_input;
    }

    /** 
     * Print the Section text
     */
    public function print_section_info()
    {
        print '
<p>If you would like to style your map, you will need call the Google map via JavaScript 
rather than embedding it.</p>
<p>This requires an <a href="https://developers.google.com/maps/documentation/javascript/">API key</a> (enable both the JavaScript API and the Static Maps API), and a valid JSON string of map options.</p>
<p><a href="https://snazzymaps.com">Snazzy Maps</a> offers a nice interface for figgerin’ out your map style.</p>
';
    }

    /** 
     * Get the settings option array and print one of its values
     */
    public function map_api_callback()
    {
        printf(
            '<input id="map_api_key" type="text" name="vamap_style[map_api_key]" value="%s">',
            isset( $this->options['map_api_key'] ) ? esc_attr( $this->options['map_api_key'] ) : ''
        );
    }

    /** 
     * Print style options for Vanilla Maps. Should be JSON.
     */
    public function map_style_callback()
    {
      $map_style = $this->options['map_style'];
      
      if( isset( $map_style ) && ! $this->is_json( $map_style) ) {
        print("<div class=\"error\"> <p>Looks like the Map Style is not valid JSON.</p></div>");
      }
      
      printf(
          '<textarea id="map_style" name="vamap_style[map_style]">%s</textarea>',
          isset( $map_style ) ? esc_attr( $map_style ) : ''
      );
    }

    /** 
     * Upload an image for a marker.
     * Cribbed from http://hughlashbrooke.com/2014/02/26/complete-versatile-options-page-class-wordpress-plugin/
     */
    public function map_icon_callback()
    {      
				$html = '';
				$data = '';
				$image_thumb = '';
				$option_name = 'map_icon';
				if( isset( $this->options['map_icon'] ) ) {
				  $data = esc_attr( $this->options['map_icon'] );
					$image_thumb = wp_get_attachment_thumb_url( $data );
				}
				
				$html .= '<img id="' . $option_name . '_preview" class="image_preview" src="' . $image_thumb . '" /><br/>' . "\n";
				$html .= '<input id="' . $option_name . '_button" type="button" data-uploader_title="' . __( 'Upload an image' , 'plugin_textdomain' ) . '" data-uploader_button_text="' . __( 'Use image' , 'plugin_textdomain' ) . '" class="image_upload_button button" value="'. __( 'Upload new image' , 'plugin_textdomain' ) . '" />' . "\n";
				$html .= '<input id="' . $option_name . '_delete" type="button" class="image_delete_button button" value="'. __( 'Remove image' , 'plugin_textdomain' ) . '" />' . "\n";
				$html .= '<input id="' . $option_name . '" class="image_data_field" type="hidden" name="vamap_style[' . $option_name . ']" value="' . $data . '"/><br/>' . "\n";
      
      echo $html;
    }
    
    /**
     * Ultra basic check for valid JSON string.
     * http://stackoverflow.com/a/6041773
     */
    function is_json($string) {
      json_decode($string);
      return (json_last_error() == JSON_ERROR_NONE);
    }
}

if( is_admin() ) {
  $my_settings_page = new VAMSettingsPage();
  
  $page = 'vamaps-options';

  /* Using registered $page handle to hook script load */
  add_action('admin_enqueue_scripts-' . $page, array($my_settings_page, 'register_map_options_scripts') );
}

js/vanilla_adaptive_maps.js
/*
 * buildMap(amap)
 * What map are we going to build?
 */
function buildMap(amap) {

  // Check if the window is wider than our setting for static maps.
  // If it is, we'll be serving up a dynamic map.
  if ( document.body.clientWidth > amap.bp ) {

    // If we have style settings saved, then we will build the map
    // natively using Google's JavaScript library so we can style it.
    if ( amap.hasOwnProperty('style') && window.hasOwnProperty('google') ) {

      // Check to see we haven't built a map already
      if ( document.getElementById(amap.id + 'js') == null ) {
        buildJS(amap);
      }

    } else { // Otherwise we will embed an iframe map.

      // Check to see we haven't built a map already
      if ( document.getElementById(amap.id + 'embed') == null ) {
        buildEmbed(amap);
      }
    }

  // If the window is narrower than our setting, get a static map
  } else {

    // Check to see we haven't built a map already
    if ( document.getElementById(amap.id + 'static') == null ) {
      buildStatic(amap);
    }
  }
}

/* geocoding from google
https://developers.google.com/maps/documentation/javascript/geocoding
 var geocoder;
  var map;
  function initialize() {
    geocoder = new google.maps.Geocoder();
    var latlng = new google.maps.LatLng(-34.397, 150.644);
    var mapOptions = {
      zoom: 8,
      center: latlng
    }
    map = new google.maps.Map(document.getElementById("map"), mapOptions);
  }

 function codeAddress() {
    var address = document.getElementById("address").value;
    geocoder.geocode( { 'address': address}, function(results, status) {
      if (status == google.maps.GeocoderStatus.OK) {
        map.setCenter(results[0].geometry.location);
        var marker = new google.maps.Marker({
            map: map,
            position: results[0].geometry.location
        });
      } else {
        alert("Geocode was not successful for the following reason: " + status);
      }
    });
  }
*/

/*
 * buildJS(amap)
 * Builds a map using Google's API directly on the page
 */
function buildJS(amap) {
  var el = document.createElement('div'),
    map = document.getElementById('map' + amap.id),
    geocoder = new google.maps.Geocoder(),
    gmap;

    el.classList.add('amap-container');
    el.setAttribute('id', amap.id + 'js');

    geocoder.geocode( { 'address': amap.addr }, function( results, status ) {

      if (status == google.maps.GeocoderStatus.OK) {
      
        var mapOptions = {
          zoom : amap.zoom,
          center : results[0].geometry.location,
          styles : amap.style,
          backgroundColor: "none"
        };
        
        console.log ( amap.style );

        gmap = new google.maps.Map(el, mapOptions);
        var marker = new google.maps.Marker({
            map: gmap,
            position: results[0].geometry.location,
            icon: amap.icon
        });
        
        console.log ( gmap );

      } else {
        alert("Geocode was not successful for the following reason: " + status);
      }

  });

  map.insertBefore(el, map.firstChild);
}

/*
 * buildEmbed(amap)
 * insert a map embedded in an iframe
 */
function buildEmbed(amap) {
  var el = document.createElement('div'),
    map = document.getElementById('map' + amap.id);

  el.classList.add('amap-container');
  el.setAttribute('id', amap.id + 'embed');
  el.innerHTML = amap.embedMap;

  map.insertBefore(el, map.firstChild);
}

/*
 * buildStatic(amap)
 * create an image based on Google's static maps API
 */
function buildStatic(amap) {
  var linkid = 'maplink' + amap.id,
    mapLink = document.getElementById(linkid).getAttribute('href'),
    img = '<img class="static-img" id="' + amap.id + 'static" src="' + amap.staticMap + '">',
    el = document.createElement('a'),
    map = document.getElementById('map' + amap.id);
    
  el.setAttribute('href', mapLink);
  el.innerHTML = img;
  
  map.insertBefore(el, map.firstChild);
}


vanilla-adaptive-maps/vanilla_adaptive_maps.php
<?php

/**
 * The plugin bootstrap file
 *
 *
 * @link              http://jeremycarlson.com/
 * @since             1.0.0
 * @package           vanilla_adaptive_maps
 *
 * @wordpress-plugin
 * Plugin Name:       Vanilla Adaptive Maps
 * Plugin URI:        http://jeremycarlson.com/
 * Description:       Include an adaptive map based on a street address with a simple shortcode. No JavaScript library required. Based on Brad Frost’s <a href="http://bradfrostweb.com/blog/post/adaptive-maps">adaptive map pattern</a>
 * Version:           1.0.0
 * Author:            Jeremy Carlson
 * Author URI:        http://jeremycarlson.com/
 * License:           GPL-2.0+
 * License URI:       http://www.gnu.org/licenses/gpl-2.0.txt
 * Text Domain:       vanilla_adaptive_maps
 * Domain Path:       /languages
 */

// If this file is called directly, abort.
if ( ! defined( 'WPINC' ) ) {
  die;
}

class vanilla_adaptive_maps {

  function set_breakpoint( $units = false ) {
    return '550' . $units;
  }

  /*
   * Adaptive Map
   * Generate the HTML from the shortcode [vamap addr="Your Street Address"]
   */
  function map_shortcode( $atts ) {

    // Encode address for mapping. This may need to get beefier.
    $map_address_encoded = str_replace(' ', '+', $atts['addr']);

    // Set up unique ID for the map, so you can have more than one on a page.
    $mid = uniqid();
    $map_id = "map$mid";
    $link_id = "maplink$mid";
    
    $output = '
<div class="adaptive-map" id="' . $map_id . '">
<a href="https://maps.apple.com/maps?q=' . $map_address_encoded . '" class="map-link" id="' . $link_id . '">' . __('View Map', 'wordpress') . '</a>
';

    // Script is distinct to this ID, and must be output inline.
    $output .= vanilla_adaptive_maps::print_map_script($map_address_encoded, $mid);
    
    // For now, we are inlining CSS as well. I realize this is not "the WordPress Way"
    // but this reduces the requests, and the CSS may eventually be something the user
    // can tweak. In particular, it would be nice to be able to adjust the breakpoints.
    $output .= vanilla_adaptive_maps::print_map_styles();

    $output .= '
</div><!-- adaptive map -->
';
    
    wp_enqueue_script( 'vanilla-adaptive-maps' );
    return $output;
  }


  /*
   * Print Map Script
   * Output the individual script for your particular map shortcode
   */
  function print_map_script($map_address_encoded, $map_id) {
  
    // get breakpoint for mobile vs desktop
    $breakpoint = vanilla_adaptive_maps::set_breakpoint();

    $script_output = "

<script>
  document.addEventListener( 'DOMContentLoaded', buildMap$map_id );
  window.addEventListener( 'resize', buildMap$map_id, false );

  function buildMap$map_id() {
  var address = '$map_address_encoded',
    staticSize = '640x320',
    amap$map_id = {
      id  : '$map_id',
      bp : $breakpoint,
      staticMap : 'http://maps.google.com/maps/api/staticmap?center=' + address + '&markers=' + address + '&size=' + staticSize + '&sensor=true',
      embedMap : '<iframe width=\"980\" height=\"650\" frameborder=\"0\" scrolling=\"no\" marginheight=\"0\" marginwidth=\"0\" src=\"https://maps.google.com/maps?q=' + address + '&output=embed\"></iframe>',
    };

    buildMap(amap$map_id);
  }
</script>

";

    return $script_output;
  }

  /*
   * Print Map Styles
   * Output the CSS for the map
   */
  function print_map_styles() {
    // get breakpoint for mobile vs desktop
    $breakpoint = vanilla_adaptive_maps::set_breakpoint('px');

    $css = "

<style type='text/css'>
.static-img {
  display: block;
}

.adaptive-map iframe {
  max-width: 100%;
}

/* From http://codepen.io/chriscoyier/full/kycDp */
.amap-container {
  width: 100%;
  margin: 0 auto;
  height: 0;
  padding-top: 38%;
  position: relative;
  display: none;
  /* Hide for small screens */
}
.amap-container iframe {
  width: 100%;
  height: 100%;
  /* had to specify height/width */
  position: absolute;
  top: 0;
  right: 0;
  left: 0;
  bottom: 0;
}

/* Medium Screens */
@media all and (min-width: $breakpoint) {
  .amap-container {
    display: block;
  }

  .static-img {
    display: none;
  }
}
</style>

";

    return $css;
  }

  // http://mikejolley.com/2013/12/02/sensible-script-enqueuing-shortcodes/
  function register_map_script() {
    wp_register_script( 'vanilla-adaptive-maps', plugins_url( '/js/vanilla_adaptive_maps.js' , __FILE__ ), array(), '1.0.0', true );
  }

} // end class

$adaptiveMap = new vanilla_adaptive_maps();
add_shortcode( 'vamap', array($adaptiveMap, 'map_shortcode') );
add_action( 'wp_enqueue_scripts', array($adaptiveMap, 'register_map_script') );


vanilla-adaptive-maps.php
<?php

/*
Plugin Name: Vanilla Adaptive Maps
Plugin URI: https://github.com/eyesofjeremy/vanilla-adaptive-maps/
Description: Include an adaptive map based on a street address with a simple shortcode. No JavaScript library required.
Version: 1.1
Author: Jeremy Carlson
Author http://jeremycarlson.com/
License: GPL-2.0+
*/

// If this file is called directly, abort.
if ( ! defined( 'WPINC' ) ) {
  die;
}

class vanilla_adaptive_maps {

  /*
   * Set breakpoint
   * TODO: make this accessible via the options page.
   * I'd like to think about that one more carefully though.
   */
  function set_breakpoint( $units = false ) {
    return '550' . $units;
  }

  /*
   * Adaptive Map
   * Generate the HTML from the shortcode [vamap addr="Your Street Address"]
   */
  function map_shortcode( $atts ) {

    $map_address = $atts['addr'];

    // Encode address for mapping. This may need to get beefier.
    $map_address_encoded = str_replace(' ', '+', $map_address);

    // Set up unique ID for the map, so you can have more than one on a page.
    $mid = uniqid();
    $map_id = "map$mid";
    $link_id = "maplink$mid";
    
    $output = '
<div class="adaptive-map" id="' . $map_id . '">
<a href="https://maps.apple.com/maps?q=' . $map_address_encoded . '" class="map-link" id="' . $link_id . '">' . __('View Map', 'wordpress') . '</a>
';

    // Script is distinct to this ID, and must be output inline.
    $output .= vanilla_adaptive_maps::print_map_script($map_address, $map_address_encoded, $mid);
    
    // For now, we are inlining CSS as well. I realize this is not "the WordPress Way"
    // but this reduces the requests, and the CSS may eventually be something the user
    // can tweak. In particular, it would be nice to be able to adjust the breakpoints.
    $output .= vanilla_adaptive_maps::print_map_styles();

    // Add a link to the options page. Make your life easier.
    if( current_user_can( 'manage_options') ) {
      $output .= '<a href="' . get_admin_url( null, 'options-general.php?page=vamap-setting-admin' ) . '" class="post-edit-link">Map Settings</a>';
    }

    $output .= '
</div><!-- adaptive map -->
';
    
    // Okay, let's do this! Get the JS functions for building the map
    wp_enqueue_script( 'vanilla-adaptive-maps' );

    // If we need the JavaScript API, let's go get 'er.
    // The check returns all options.
    if( vanilla_adaptive_maps::use_jsapi() ) {
      wp_enqueue_script( 'vanilla-adaptive-maps-google-maps' );
    }

    return $output;
  }
  
  /*
   * Check if we need the JavaScript API
   */
  function use_jsapi() {
  
    $options = get_option('vamap_style');
    
    return ( isset( $options['map_api_key'] ) ) ? $options : FALSE;
  }

  /*
   * Print Map Script
   * Output the individual script for your particular map shortcode
   */
  function print_map_script($map_address, $map_address_encoded, $map_id) {
    
    $options = get_option('vamap_style');

    $map_style = '[]'; // empty array for JS map style
    $static_style = ''; // empty string for static map style
    if( isset( $options['map_style'] ) && ! empty( $options['map_style'] ) ) {
      $static_style = vanilla_adaptive_maps::url_args_from_style_json( $options['map_style'] );
      
      if( isset( $options['map_api_key']) ) {
        $map_style = $options['map_style'];
      }
    }
    
    if( isset( $options['map_icon'] ) ) {
      $map_icon = wp_get_attachment_url( $options['map_icon'] );
      $static_icon = "icon:$map_icon|";
    } else {
      $map_icon = '';
      $static_icon = '';
    }
    

    // get breakpoint for mobile vs desktop
    $breakpoint = vanilla_adaptive_maps::set_breakpoint();

    $script_output = "

<script>
  document.addEventListener( 'DOMContentLoaded', buildMap$map_id );
  window.addEventListener( 'resize', buildMap$map_id, false );

  function buildMap$map_id() {
  var address = '$map_address_encoded',
    staticSize = '640x320',
    amap$map_id = {
      id  : '$map_id',
      bp : $breakpoint,
      addr : '$map_address_encoded',
      zoom : 14,
      icon : '$map_icon',
      style : $map_style,
      staticMap : 'http://maps.google.com/maps/api/staticmap?center=' + address + '&markers=$static_icon' + address + '&size=' + staticSize + '&style=' + '$static_style' + '&sensor=true',
      embedMap : '<iframe title=\"Google map to $map_address\" width=\"980\" height=\"650\" frameborder=\"0\" scrolling=\"no\" marginheight=\"0\" marginwidth=\"0\" src=\"https://maps.google.com/maps?q=' + address + '&output=embed\"></iframe>',
    };

    buildMap(amap$map_id);
  }
</script>

";

    return $script_output;
  }

  /*
   * Print Map Styles
   * Output the CSS for the map
   */
  function print_map_styles() {
    // get breakpoint for mobile vs desktop
    $breakpoint = vanilla_adaptive_maps::set_breakpoint('px');

    $map_loading = __('Map Loading', 'vanilla-adaptive-maps');
    $css = "

<style type='text/css'>
.adaptive-map, .static-img {
  display: block;
  position: relative;
}

.adaptive-map iframe {
  max-width: 100%;
}

.adaptive-map:before {
content: '" . $map_loading . "';
  display: block;
  position: absolute;
  text-align: center;
  width: 12em;
  margin-left: -6em;
  margin-top: -.5em;
  top: 50%;
  left: 50%;
  animation: blink 2s linear 0s infinite;
  -webkit-animation: blink 2s linear 0s infinite;
}

@keyframes blink {
  50% {
    opacity: 0.0;
  }
}
@-webkit-keyframes blink {
  50% {
    opacity: 0.0;
  }
}

/* From http://codepen.io/chriscoyier/full/kycDp */
.amap-container {
  width: 100%;
  margin: 0 auto;
  height: 0;
  padding-top: 38%;
  position: relative;
  display: none;
  /* Hide for small screens */
}
.amap-container iframe {
  width: 100%;
  height: 100%;
  /* had to specify height/width */
  position: absolute;
  top: 0;
  right: 0;
  left: 0;
  bottom: 0;
}

/* Medium Screens */
@media all and (min-width: $breakpoint) {
  .amap-container {
    display: block;
  }

  .static-img {
    display: none;
  }
}
</style>

";

    return $css;
  }
  
  // http://stackoverflow.com/a/28173229/156645
  function url_args_from_style_json($mapStyleJson) {
    $params = [];
    
    $json_object = json_decode($mapStyleJson, true);

    foreach ($json_object as $style) {
      $styleString = '';

      if (isset($style['stylers']) && count($style['stylers']) > 0) {
        $styleString .= (isset($style['featureType']) ? ('feature:' . $style['featureType']) : 'feature:all') . '|';
        $styleString .= (isset($style['elementType']) ? ('element:' . $style['elementType']) : 'element:all') . '|';

        foreach ($style['stylers'] as $styler) {
          $propertyname = array_keys($styler)[0];
          $propertyval = str_replace('#', '0x', $styler[$propertyname]);
          $styleString .= $propertyname . ':' . $propertyval . '|';
        }
      }

      $styleString = substr($styleString, 0, strlen($styleString) - 1);

      $params[] = 'style=' . $styleString;
    }

    return implode("&", $params);
  }

  // http://mikejolley.com/2013/12/02/sensible-script-enqueuing-shortcodes/
  function register_map_script() {
    wp_register_script(
      'vanilla-adaptive-maps', // script hook
      plugins_url( '/js/vanilla_adaptive_maps.js' , __FILE__ ), // script location
      array(), // dependencies
      '1.0.0', // script version
      true // in footer?
    );
    
    // If we need the JavaScript API, let's go get 'er.
    // The check returns all options.
    $use_api = vanilla_adaptive_maps::use_jsapi();
    
    if( $use_api ) {
      $api_key = $use_api['map_api_key'];

      wp_register_script(
        'vanilla-adaptive-maps-google-maps', // script hook
        'https://maps.googleapis.com/maps/api/js?v=3.exp&libraries=places&key=' . $api_key, // script location
        array(), // dependencies
        '1.0.0', // script version
        true // in footer?
      );
    }
  }

} // end class

$adaptiveMap = new vanilla_adaptive_maps();
add_shortcode( 'vamap', array($adaptiveMap, 'map_shortcode') );
add_action( 'wp_enqueue_scripts', array($adaptiveMap, 'register_map_script') );

// Add options for the plugin
include('vamaps-options.php');